services/logging/src/test/logging.controller.test.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright IBM Corp. 2021, 2021
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import { LogLoggedMessage } from '@carbon-platform/api/logging'
+
+import { LogDnaService } from '../main/log-dna.service'
+import { LoggingController } from '../main/logging.controller'
+
+test('logLogged runs without crashing', async () => {
+  const logDnaService = new LogDnaService()
+  const loggingController = new LoggingController(logDnaService)
+
+  const data: LogLoggedMessage = {
+    component: 'test',
+    environment: 'test',
+    level: 'debug',
+    message: 'test message',
+    service: 'test service',
+    timestamp: Date.now()
+  }
+
+  await expect(loggingController.logLogged(data)).resolves.not.toThrow()
+})
+
+describe('message validation', () => {
+  let logDnaService: LogDnaService
+  let loggingController: LoggingController
+  let data: LogLoggedMessage
+
+  beforeEach(() => {
+    logDnaService = new LogDnaService()
+    loggingController = new LoggingController(logDnaService)
+    ;(loggingController as any).nestLogger = {
+      warn: jest.fn()
+    }
+
+    data = {
+      component: 'test',
+      environment: 'test',
+      level: 'debug',
+      message: 'test message',
+      service: 'test service',
+      timestamp: Date.now()
+    }
+  })
+
+  it('logs a warning when no component specified', async () => {
+    delete (data as any).component
```

Comment:
why do we need to cast to any?

Reply:
None of the fields on `data` are marked as optional (and they're all required in the map passed to `logLogged()`, so to get around this and fake an incoming request that is missing one of the fields, I first assert it as `any`, which gives unmitigated access to the object properties, delete one of the required fields and the pass it in.

As an alternative to this, I could instead have two interfaces. One called `LogLoggedMessage` and another called like `UnvalidatedLogLoggedMessage`, and only begin using `LogLoggedMessage` once validation has been performed successfully.

What are your thoughts on that as a general paradigm?

Reply:
oh ok I understand now. I don't think having the two interfaces is necessary we can leave like this unless you really want to change it

Reply:
I think I am gonna do two actually haha. I think this would be a good case for using type assertions on methods like you and I chatted about the other day. So the validator function would become like:

```ts
function validate(data: UnvalidatedMessage): data is LogLoggedMessage {
  // do all the checks to ensure data is a valid LogLoggedMessage
  // and `throw` otherwise
}
```

I just need to wait until #192 is merged so I can have access to LogLoggedMessage on this branch.